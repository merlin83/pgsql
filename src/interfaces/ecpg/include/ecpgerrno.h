#ifndef _ECPG_ERROR_H
#define _ECPG_ERROR_H

#include <errno.h>

/* This is a list of all error codes the embedded SQL program can return */
#define	ECPG_NO_ERROR		0
#define ECPG_NOT_FOUND		100

/* system error codes returned by ecpglib get the correct number,
 * but are made negative
 */
#define ECPG_OUT_OF_MEMORY	-ENOMEM

/* first we have a set of ecpg messages, they start at 200 */
#define ECPG_UNSUPPORTED	-200
#define ECPG_TOO_MANY_ARGUMENTS	-201
#define ECPG_TOO_FEW_ARGUMENTS	-202
#define ECPG_TOO_MANY_MATCHES	-203
#define ECPG_INT_FORMAT		-204
#define ECPG_UINT_FORMAT	-205
#define ECPG_FLOAT_FORMAT	-206
#define ECPG_CONVERT_BOOL	-207
#define ECPG_EMPTY		-208
#define ECPG_NO_CONN		-209

/* finally the backend error messages, they start at 300 */
#define ECPG_PGSQL		-300
#define ECPG_TRANS		-301
#define ECPG_CONNECT		-302

#endif /* !_ECPG_ERROR_H */
