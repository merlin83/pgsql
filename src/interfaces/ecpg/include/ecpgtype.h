/*
 * This file implements a data structure that is built and maintained by the
 * preprocessor.
 *
 * All types that can be handled for host variable declarations has to
 * be handled eventually.
 */

/*
 * Here are all the types that we are to handle. Note that it is the type
 * that is registered and that has nothing whatsoever to do with the storage
 * class.
 *
 * Simle types
 * integers: char, short, int, long (signed and unsigned)
 * floats: float, double
 *
 * Complex types:
 * VARCHAR, VARCHAR2 - Strings with length (maxlen is given in the declaration)
 * Arrays of simple types and of VARCHAR, VARCHAR2 (size given in declaration)
 * Records build of simple types, arrays and other records.
 *
 * Complicating things:
 * typedefs and struct names!
 *
 * Conclusion:
 * This is a typically recursive definition. A structure of typed list elements
 * would probably work fine:
 */
#include <stdio.h>

enum ECPGttype {
    ECPGt_char = 1, ECPGt_unsigned_char, ECPGt_short, ECPGt_unsigned_short, 
    ECPGt_int, ECPGt_unsigned_int, ECPGt_long, ECPGt_unsigned_long,
    ECPGt_bool,
    ECPGt_float, ECPGt_double,
    ECPGt_varchar, ECPGt_varchar2,
    ECPGt_array,
    ECPGt_record,
    ECPGt_EOIT,			/* End of insert types. */
    ECPGt_EORT			/* End of result types. */
};

#define IS_SIMPLE_TYPE(type) ((type) >= ECPGt_char && (type) <= ECPGt_varchar2)
