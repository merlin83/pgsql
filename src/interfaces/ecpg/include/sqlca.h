#ifndef POSTGRES_SQLCA_H
#define POSTGRES_SQLCA_H

#ifdef __cplusplus
extern "C" {
#endif

struct sqlca
{
	int			sqlcode;
	struct
	{
		int			sqlerrml;
		char		sqlerrmc[1000];
	}			sqlerrm;
}			sqlca;

#endif

#ifdef __cplusplus
}
#endif
 
