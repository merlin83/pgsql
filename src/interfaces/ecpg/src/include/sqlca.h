#ifndef POSTGRES_SQLCA_H
#define POSTGRES_SQLCA_H

struct sqlca {
    int sqlcode;
    struct {
	int sqlerrml;
	char sqlerrmc[1000];
    } sqlerrm;
} sqlca;
#endif
