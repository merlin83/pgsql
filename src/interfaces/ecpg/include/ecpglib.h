#include <postgres.h>

#ifdef __cplusplus
extern		"C"
{
#endif

	void		ECPGdebug(int, FILE *);
	bool		ECPGstatus(int, const char *);
	bool		ECPGsetconn(int, const char *);
	bool		ECPGconnect(int, const char *, const char *, const char *, const char *);
	bool		ECPGdo(int, const char *, char *,...);
	bool		ECPGtrans(int, const char *, const char *);
	bool		ECPGdisconnect(int, const char *);
	bool		ECPGprepare(int, char *, char *);
	bool		ECPGdeallocate(int, char *);
	char		*ECPGprepared_statement(char *);
		
	void		ECPGlog(const char *format,...);

#ifdef LIBPQ_FE_H
	bool		ECPGsetdb(PGconn *);

#endif

/* Here are some methods used by the lib. */
/* Returns a pointer to a string containing a simple type name. */
	const char *ECPGtype_name(enum ECPGttype);

/* A generic varchar type. */
	struct ECPGgeneric_varchar
	{
		int			len;
		char		arr[1];
	};

/* print an error message */
	void		sqlprint(void);

	struct cursor
	{
		const char *name;
		char	   *command;
		struct cursor *next;
	};

	extern int	no_auto_trans;

/* define this for simplicity as well as compatibility */

#define		  SQLCODE	 sqlca.sqlcode

#ifdef __cplusplus
}

#endif

#include <ecpgerrno.h>
