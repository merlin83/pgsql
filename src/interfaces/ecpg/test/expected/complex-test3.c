/* Processed by ecpg (4.2.1) */
/* These include files are added by the preprocessor */
#include <ecpgtype.h>
#include <ecpglib.h>
#include <ecpgerrno.h>
#include <sqlca.h>
/* End of automatic include section */

#line 1 "test3.pgc"
/****************************************************************************/
/* Test comment                                                             */
/*--------------------------------------------------------------------------*/

#line 1 "./header_test.h"
#include "stdlib.h"

static void
Finish(char *msg)
{
	fprintf(stderr, "Error in statement '%s':\n", msg);
	sqlprint();

	/* finish transaction */
	{ ECPGtrans(__LINE__, NULL, "rollback");}
#line 10 "./header_test.h"


	/* and remove test table */
	{ ECPGdo(__LINE__, 0, 1, NULL, "drop table meskes ", ECPGt_EOIT, ECPGt_EORT);}
#line 13 "./header_test.h"

	{ ECPGtrans(__LINE__, NULL, "commit");}
#line 14 "./header_test.h"


	{ ECPGdisconnect(__LINE__, "CURRENT");}
#line 16 "./header_test.h"


	exit(-1);
}

static void
warn(void)
{
	fprintf(stderr, "Warning: At least one column was truncated\n");
}

/* exec sql whenever sqlerror  do Finish ( msg ) ; */
#line 29 "./header_test.h"

/* exec sql whenever sql_warning  do warn (  ) ; */
#line 32 "./header_test.h"


#line 4 "test3.pgc"


#line 1 "./../regression.h"






#line 5 "test3.pgc"


/* exec sql type str is  [ 10 ]   */
#line 7 "test3.pgc"


#include <stdlib.h>
#include <string.h>

int
main (void)
{
/* exec sql begin declare section */
	        typedef struct { 
#line 16 "test3.pgc"
 long  born    ;
 
#line 16 "test3.pgc"
 short  age    ;
 }  birthinfo ;

#line 16 "test3.pgc"

	 		 
					 
				 
	  	 
					 
				   
	   
	    
	 
	   
	 
	 
 
#line 19 "test3.pgc"
 struct personal_struct { 
#line 17 "test3.pgc"
   struct varchar_name  { int len; char arr[ 10 ]; }  name    ;
 
#line 18 "test3.pgc"
 birthinfo  birth    ;
 }  personal    ;
 
#line 22 "test3.pgc"
 struct personal_indicator { 
#line 20 "test3.pgc"
 int  ind_name    ;
 
#line 21 "test3.pgc"
 birthinfo  ind_birth    ;
 }  ind_personal    ;
 
#line 23 "test3.pgc"
 int * ind_married   = NULL ;
 
#line 24 "test3.pgc"
 int  children    ,  movevalue   = 2 ;
 
#line 25 "test3.pgc"
 int  ind_children    ;
 
#line 26 "test3.pgc"
   struct varchar_married  { int len; char arr[ 10 ]; } * married  = NULL ;
 
#line 27 "test3.pgc"
 char * wifesname   = "Petra" ;
 
#line 28 "test3.pgc"
 char * query   = "select * from meskes where name = ?" ;
/* exec sql end declare section */
#line 29 "test3.pgc"


	/* declare cur  cursor  for select  name , born , age , married , children  from meskes    */
#line 32 "test3.pgc"


	char msg[128];

	ECPGdebug(1, stderr);

	strcpy(msg, "connect");
	{ ECPGconnect(__LINE__, 0, "regress1" , NULL,NULL , NULL, 0); 
#line 39 "test3.pgc"

if (sqlca.sqlwarn[0] == 'W') warn (  );
#line 39 "test3.pgc"

if (sqlca.sqlcode < 0) Finish ( msg );}
#line 39 "test3.pgc"


	strcpy(msg, "create");
	{ ECPGdo(__LINE__, 0, 1, NULL, "create  table meskes ( name char  ( 8 )    , born integer   , age smallint   , married date   , children integer   )    ", ECPGt_EOIT, ECPGt_EORT);
#line 42 "test3.pgc"

if (sqlca.sqlwarn[0] == 'W') warn (  );
#line 42 "test3.pgc"

if (sqlca.sqlcode < 0) Finish ( msg );}
#line 42 "test3.pgc"


	strcpy(msg, "insert");
	{ ECPGdo(__LINE__, 0, 1, NULL, "insert into meskes ( name  , married  , children  ) values(  ? , '19900404' , 3 )", 
	ECPGt_char,&(wifesname),(long)0,(long)1,(1)*sizeof(char), 
	ECPGt_NO_INDICATOR, NULL , 0L, 0L, 0L, ECPGt_EOIT, ECPGt_EORT);
#line 45 "test3.pgc"

if (sqlca.sqlwarn[0] == 'W') warn (  );
#line 45 "test3.pgc"

if (sqlca.sqlcode < 0) Finish ( msg );}
#line 45 "test3.pgc"

	{ ECPGdo(__LINE__, 0, 1, NULL, "insert into meskes ( name  , born  , age  , married  , children  ) values( 'Michael' , 19660117 , 35 , '19900404' , 3 )", ECPGt_EOIT, ECPGt_EORT);
#line 46 "test3.pgc"

if (sqlca.sqlwarn[0] == 'W') warn (  );
#line 46 "test3.pgc"

if (sqlca.sqlcode < 0) Finish ( msg );}
#line 46 "test3.pgc"

	{ ECPGdo(__LINE__, 0, 1, NULL, "insert into meskes ( name  , born  , age  ) values( 'Carsten' , 19910103 , 10 )", ECPGt_EOIT, ECPGt_EORT);
#line 47 "test3.pgc"

if (sqlca.sqlwarn[0] == 'W') warn (  );
#line 47 "test3.pgc"

if (sqlca.sqlcode < 0) Finish ( msg );}
#line 47 "test3.pgc"

	{ ECPGdo(__LINE__, 0, 1, NULL, "insert into meskes ( name  , born  , age  ) values( 'Marc' , 19930907 , 8 )", ECPGt_EOIT, ECPGt_EORT);
#line 48 "test3.pgc"

if (sqlca.sqlwarn[0] == 'W') warn (  );
#line 48 "test3.pgc"

if (sqlca.sqlcode < 0) Finish ( msg );}
#line 48 "test3.pgc"

	{ ECPGdo(__LINE__, 0, 1, NULL, "insert into meskes ( name  , born  , age  ) values( 'Chris' , 19970923 , 4 )", ECPGt_EOIT, ECPGt_EORT);
#line 49 "test3.pgc"

if (sqlca.sqlwarn[0] == 'W') warn (  );
#line 49 "test3.pgc"

if (sqlca.sqlcode < 0) Finish ( msg );}
#line 49 "test3.pgc"


	strcpy(msg, "commit");
	{ ECPGtrans(__LINE__, NULL, "commit");
#line 52 "test3.pgc"

if (sqlca.sqlwarn[0] == 'W') warn (  );
#line 52 "test3.pgc"

if (sqlca.sqlcode < 0) Finish ( msg );}
#line 52 "test3.pgc"


	strcpy(msg, "open");
	{ ECPGdo(__LINE__, 0, 1, NULL, "declare cur  cursor  for select  name , born , age , married , children  from meskes   ", ECPGt_EOIT, ECPGt_EORT);
#line 55 "test3.pgc"

if (sqlca.sqlwarn[0] == 'W') warn (  );
#line 55 "test3.pgc"

if (sqlca.sqlcode < 0) Finish ( msg );}
#line 55 "test3.pgc"


	strcpy(msg, "move");
	{ ECPGdo(__LINE__, 0, 1, NULL, "move  ? in cur", 
	ECPGt_int,&(movevalue),(long)1,(long)1,sizeof(int), 
	ECPGt_NO_INDICATOR, NULL , 0L, 0L, 0L, ECPGt_EOIT, ECPGt_EORT);
#line 58 "test3.pgc"

if (sqlca.sqlwarn[0] == 'W') warn (  );
#line 58 "test3.pgc"

if (sqlca.sqlcode < 0) Finish ( msg );}
#line 58 "test3.pgc"


	/* exec sql whenever not found  break ; */
#line 60 "test3.pgc"


	while (1) {
		strcpy(msg, "fetch");
		{ ECPGdo(__LINE__, 0, 1, NULL, "fetch from cur", ECPGt_EOIT, 
	ECPGt_varchar,&(personal.name),(long)10,(long)1,sizeof(struct varchar_name), 
	ECPGt_int,&(ind_personal.ind_name),(long)1,(long)1,sizeof(int), 
	ECPGt_long,&(personal.birth.born),(long)1,(long)1,sizeof(long), 
	ECPGt_long,&(ind_personal.ind_birth.born),(long)1,(long)1,sizeof(long), 
	ECPGt_short,&(personal.birth.age),(long)1,(long)1,sizeof(short), 
	ECPGt_short,&(ind_personal.ind_birth.age),(long)1,(long)1,sizeof(short), 
	ECPGt_varchar,&(married),(long)10,(long)0,sizeof(struct varchar_married), 
	ECPGt_int,&(ind_married),(long)1,(long)0,sizeof(int), 
	ECPGt_int,&(children),(long)1,(long)1,sizeof(int), 
	ECPGt_int,&(ind_children),(long)1,(long)1,sizeof(int), ECPGt_EORT);
#line 64 "test3.pgc"

if (sqlca.sqlcode == ECPG_NOT_FOUND) break;
#line 64 "test3.pgc"

if (sqlca.sqlwarn[0] == 'W') warn (  );
#line 64 "test3.pgc"

if (sqlca.sqlcode < 0) Finish ( msg );}
#line 64 "test3.pgc"

		printf("%8.8s", personal.name.arr);
		if (ind_personal.ind_birth.born >= 0)
			printf(", born %ld", personal.birth.born);
		if (ind_personal.ind_birth.age >= 0)
			printf(", age = %d", personal.birth.age);
		if (*ind_married >= 0)
			printf(", married %10.10s", married->arr);
		if (ind_children >= 0)
			printf(", children = %d", children);
		putchar('\n');

		free(married);
		married = NULL;
	}

	strcpy(msg, "close");
	{ ECPGdo(__LINE__, 0, 1, NULL, "close cur", ECPGt_EOIT, ECPGt_EORT);
#line 81 "test3.pgc"

if (sqlca.sqlwarn[0] == 'W') warn (  );
#line 81 "test3.pgc"

if (sqlca.sqlcode < 0) Finish ( msg );}
#line 81 "test3.pgc"


	/* and now a query with prepare */
	{ ECPGprepare(__LINE__, "MM" , query);
#line 84 "test3.pgc"

if (sqlca.sqlwarn[0] == 'W') warn (  );
#line 84 "test3.pgc"

if (sqlca.sqlcode < 0) Finish ( msg );}
#line 84 "test3.pgc"

	/* declare prep  cursor  for ? */
#line 85 "test3.pgc"


	strcpy(msg, "open");
	{ ECPGdo(__LINE__, 0, 1, NULL, "declare prep  cursor  for ?", 
	ECPGt_char_variable,(ECPGprepared_statement("MM")),(long)1,(long)1,(1)*sizeof(char), 
	ECPGt_NO_INDICATOR, NULL , 0L, 0L, 0L, 
	ECPGt_char,&(wifesname),(long)0,(long)1,(1)*sizeof(char), 
	ECPGt_NO_INDICATOR, NULL , 0L, 0L, 0L, ECPGt_EOIT, ECPGt_EORT);
#line 88 "test3.pgc"

if (sqlca.sqlwarn[0] == 'W') warn (  );
#line 88 "test3.pgc"

if (sqlca.sqlcode < 0) Finish ( msg );}
#line 88 "test3.pgc"


	/* exec sql whenever not found  break ; */
#line 90 "test3.pgc"


	while (1) {
		strcpy(msg, "fetch");
		{ ECPGdo(__LINE__, 0, 1, NULL, "fetch in prep", ECPGt_EOIT, 
	ECPGt_varchar,&(personal.name),(long)10,(long)1,sizeof(struct varchar_name), 
	ECPGt_int,&(ind_personal.ind_name),(long)1,(long)1,sizeof(int), 
	ECPGt_long,&(personal.birth.born),(long)1,(long)1,sizeof(long), 
	ECPGt_long,&(ind_personal.ind_birth.born),(long)1,(long)1,sizeof(long), 
	ECPGt_short,&(personal.birth.age),(long)1,(long)1,sizeof(short), 
	ECPGt_short,&(ind_personal.ind_birth.age),(long)1,(long)1,sizeof(short), 
	ECPGt_varchar,&(married),(long)10,(long)0,sizeof(struct varchar_married), 
	ECPGt_int,&(ind_married),(long)1,(long)0,sizeof(int), 
	ECPGt_int,&(children),(long)1,(long)1,sizeof(int), 
	ECPGt_int,&(ind_children),(long)1,(long)1,sizeof(int), ECPGt_EORT);
#line 94 "test3.pgc"

if (sqlca.sqlcode == ECPG_NOT_FOUND) break;
#line 94 "test3.pgc"

if (sqlca.sqlwarn[0] == 'W') warn (  );
#line 94 "test3.pgc"

if (sqlca.sqlcode < 0) Finish ( msg );}
#line 94 "test3.pgc"

		printf("%8.8s", personal.name.arr);
		if (ind_personal.ind_birth.born >= 0)
			printf(", born %ld", personal.birth.born);
		if (ind_personal.ind_birth.age >= 0)
			printf(", age = %d", personal.birth.age);
		if (*ind_married >= 0)
			printf(", married %10.10s", married->arr);
		if (ind_children >= 0)
			printf(", children = %d", children);
		putchar('\n');
	}

	free(married);

	strcpy(msg, "close");
	{ ECPGdo(__LINE__, 0, 1, NULL, "close prep", ECPGt_EOIT, ECPGt_EORT);
#line 110 "test3.pgc"

if (sqlca.sqlwarn[0] == 'W') warn (  );
#line 110 "test3.pgc"

if (sqlca.sqlcode < 0) Finish ( msg );}
#line 110 "test3.pgc"


	strcpy(msg, "drop");
	{ ECPGdo(__LINE__, 0, 1, NULL, "drop table meskes ", ECPGt_EOIT, ECPGt_EORT);
#line 113 "test3.pgc"

if (sqlca.sqlwarn[0] == 'W') warn (  );
#line 113 "test3.pgc"

if (sqlca.sqlcode < 0) Finish ( msg );}
#line 113 "test3.pgc"


	strcpy(msg, "commit");
	{ ECPGtrans(__LINE__, NULL, "commit");
#line 116 "test3.pgc"

if (sqlca.sqlwarn[0] == 'W') warn (  );
#line 116 "test3.pgc"

if (sqlca.sqlcode < 0) Finish ( msg );}
#line 116 "test3.pgc"


	strcpy(msg, "disconnect"); 
	{ ECPGdisconnect(__LINE__, "CURRENT");
#line 119 "test3.pgc"

if (sqlca.sqlwarn[0] == 'W') warn (  );
#line 119 "test3.pgc"

if (sqlca.sqlcode < 0) Finish ( msg );}
#line 119 "test3.pgc"


	return (0);
}
