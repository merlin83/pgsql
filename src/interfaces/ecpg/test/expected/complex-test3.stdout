Carsten , born 19910103, age = 10
Marc    , born 19930907, age = 8
Chris   , born 19970923, age = 4
Petra   , married 04-04-1990, children = 3
